File: 130-surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        row = len(board)
        col = len(board[0])
        
        def dfs(i, j):
            if i < 0 or j < 0 or i == row or j == col or board[i][j]!='O':
                return
            board[i][j] = 'T'
            dfs(i + 1, j)
            dfs(i - 1, j)
            dfs(i, j + 1)
            dfs(i, j - 1)

        for i in range(row):
            for j in range(col):
                if (i in [0, row-1] or j in [0, col -1]) and board[i][j]:
                    dfs(i, j)

        for i in range(row):
            for j in range(col):
                if board[i][j] == 'O':
                    board[i][j] = 'X'

        for i in range(row):
            for j in range(col):
                if board[i][j] == 'T':
                    board[i][j] = 'O'

File: 130-surrounded-regions/surrounded-regions.cpp
class Solution {
public:
    int row;
    int col;

    void dfs(int i, int j, vector<vector<char>>& board){
        if (i < 0 || j < 0 || i == row || j == col || board[i][j] != 'O') return;
        board[i][j] = 'T';
        dfs(i + 1, j, board);
        dfs(i - 1, j, board);
        dfs(i, j + 1, board);
        dfs(i, j - 1, board);
    }

    void solve(vector<vector<char>>& board) {
        row = board.size();
        col = board[0].size();

        for(int i = 0; i < row; i++)
            for(int j = 0; j < col ; j++){
                if(board[i][j]=='O' and (i==0 or i == row-1 or j == 0 or j == col - 1)) dfs(i, j, board);
            }
        
        for(int i = 0; i < row; i++)
            for(int j = 0; j < col ; j++)
                if(board[i][j] == 'O')
                    board[i][j] = 'X';

        for(int i = 0; i < row; i++)
            for(int j = 0; j < col ; j++)
                if(board[i][j] == 'T')
                    board[i][j] = 'O';

    }
};